File: crates/agentpin-cli/src/issue.rs
use std::fs;

use clap::Args;

use agentpin::credential::issue_credential;
use agentpin::crypto;
use agentpin::types::capability::Capability;
use agentpin::types::constraint::Constraints;
use agentpin::types::credential::DelegationAttestation;

#[derive(Args)]
pub struct IssueArgs {
    /// Path to private key PEM file
    #[arg(long)]
    pub private_key: String,

    /// Key identifier
    #[arg(long)]
    pub kid: String,

    /// Issuer domain
    #[arg(long)]
    pub issuer: String,

    /// Agent URN (e.g., "urn:agentpin:example.com:my-agent")
    #[arg(long)]
    pub agent_id: String,

    /// Audience domain (optional)
    #[arg(long)]
    pub audience: Option<String>,

    /// Comma-separated capabilities (e.g., "read:data,write:reports")
    #[arg(long)]
    pub capabilities: String,

    /// Credential TTL in seconds
    #[arg(long, default_value = "3600")]
    pub ttl: u64,

    /// JSON file with delegation chain entries
    #[arg(long)]
    pub delegation_chain: Option<String>,

    /// JSON string or file path with constraint overrides
    #[arg(long)]
    pub constraints: Option<String>,
}

pub fn run(args: IssueArgs) -> anyhow::Result<()> {
    let pem = fs::read_to_string(&args.private_key)?;
    let signing_key = crypto::load_signing_key(&pem)?;

    let capabilities: Vec<Capability> = args
        .capabilities
        .split(',')
        .map(|s| Capability::from(s.trim()))
        .collect();

    let constraints: Option<Constraints> = match &args.constraints {
        Some(c) => {
            let json = if std::path::Path::new(c).exists() {
                fs::read_to_string(c)?
            } else {
                c.clone()
            };
            Some(serde_json::from_str(&json)?)
        }
        None => None,
    };

    let delegation_chain: Option<Vec<DelegationAttestation>> = match &args.delegation_chain {
        Some(path) => {
            let json = fs::read_to_string(path)?;
            Some(serde_json::from_str(&json)?)
        }
        None => None,
    };

    let jwt = issue_credential(
        &signing_key,
        &args.kid,
        &args.issuer,
        &args.agent_id,
        args.audience.as_deref(),
        capabilities,
        constraints,
        delegation_chain,
        args.ttl,
    )?;

    println!("{}", jwt);
    Ok(())
}

File: go/cmd/agentpin/issue.go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"os"
	"strings"

	"github.com/ThirdKeyAi/agentpin/go/pkg/credential"
	"github.com/ThirdKeyAi/agentpin/go/pkg/crypto"
	"github.com/ThirdKeyAi/agentpin/go/pkg/types"
)

func runIssue(args []string) error {
	fs := flag.NewFlagSet("issue", flag.ContinueOnError)
	privKeyPath := fs.String("private-key", "", "Path to private key PEM file (required)")
	kid := fs.String("kid", "", "Key identifier (required)")
	issuer := fs.String("issuer", "", "Issuer domain (required)")
	agentID := fs.String("agent-id", "", "Agent URN (required)")
	audience := fs.String("audience", "", "Audience domain (optional)")
	caps := fs.String("capabilities", "", "Comma-separated capabilities (required)")
	ttl := fs.Uint64("ttl", 3600, "Credential TTL in seconds")
	delChainPath := fs.String("delegation-chain", "", "JSON file with delegation chain entries")
	constraintsArg := fs.String("constraints", "", "JSON string or file with constraint overrides")
	if err := fs.Parse(args); err != nil {
		return err
	}
	if *privKeyPath == "" || *kid == "" || *issuer == "" || *agentID == "" || *caps == "" {
		return fmt.Errorf("--private-key, --kid, --issuer, --agent-id, --capabilities are required")
	}

	pemData, err := os.ReadFile(*privKeyPath)
	if err != nil {
		return err
	}
	priv, err := crypto.LoadPrivateKey(string(pemData))
	if err != nil {
		return err
	}

	capList := []types.Capability{}
	for _, c := range strings.Split(*caps, ",") {
		c = strings.TrimSpace(c)
		if c == "" {
			continue
		}
		capList = append(capList, types.Capability(c))
	}

	var constraints *types.Constraints
	if *constraintsArg != "" {
		var data []byte
		if _, err := os.Stat(*constraintsArg); err == nil {
			data, err = os.ReadFile(*constraintsArg)
			if err != nil {
				return err
			}
		} else {
			data = []byte(*constraintsArg)
		}
		var c types.Constraints
		if err := json.Unmarshal(data, &c); err != nil {
			return fmt.Errorf("invalid constraints JSON: %w", err)
		}
		constraints = &c
	}

	var chain []types.DelegationAttestation
	if *delChainPath != "" {
		data, err := os.ReadFile(*delChainPath)
		if err != nil {
			return err
		}
		if err := json.Unmarshal(data, &chain); err != nil {
			return fmt.Errorf("invalid delegation chain JSON: %w", err)
		}
	}

	jwt, err := credential.IssueCredential(priv, *kid, *issuer, *agentID, *audience, capList, constraints, chain, *ttl)
	if err != nil {
		return err
	}
	fmt.Println(jwt)
	return nil
}
